public/app/features/alerting/unified/components/rule-editor/query-and-alert-condition/QueryAndExpressionsStep.tsx

```diff
@@ -0,0 +1,252 @@
+import React, { FC, useCallback, useEffect, useMemo, useReducer, useRef, useState } from 'react';
+import { useFormContext } from 'react-hook-form';
+
+import { LoadingState, PanelData } from '@grafana/data';
+import { selectors } from '@grafana/e2e-selectors';
+import { config } from '@grafana/runtime';
+import { Alert, Button, Field, InputControl, Stack, Tooltip } from '@grafana/ui';
+import { isExpressionQuery } from 'app/features/expressions/guards';
+import { AlertQuery } from 'app/types/unified-alerting-dto';
+
+import { AlertingQueryRunner } from '../../../state/AlertingQueryRunner';
+import { RuleFormType, RuleFormValues } from '../../../types/rule-form';
+import { getDefaultOrFirstCompatibleDataSource } from '../../../utils/datasource';
+import { ExpressionEditor } from '../ExpressionEditor';
+import { ExpressionsEditor } from '../ExpressionsEditor';
+import { QueryEditor } from '../QueryEditor';
+import { RuleEditorSection } from '../RuleEditorSection';
+import { refIdExists } from '../util';
+
+import { AlertType } from './AlertType';
+import {
+  duplicateQuery,
+  addNewDataQuery,
+  addNewExpression,
+  queriesAndExpressionsReducer,
+  removeExpression,
+  rewireExpressions,
+  setDataQueries,
+  updateExpression,
+  updateExpressionRefId,
+  updateExpressionType,
+} from './reducer';
+
+interface Props {
+  editingExistingRule: boolean;
+}
+
+export const QueryAndExpressionsStep: FC<Props> = ({ editingExistingRule }) => {
+  const runner = useRef(new AlertingQueryRunner());
+  const {
+    setValue,
+    getValues,
+    watch,
+    formState: { errors },
+    control,
+  } = useFormContext<RuleFormValues>();
+  const [panelData, setPanelData] = useState<Record<string, PanelData>>({});
+
+  const initialState = {
+    queries: getValues('queries'),
+    panelData: {},
+  };
+  const [{ queries }, dispatch] = useReducer(queriesAndExpressionsReducer, initialState);
+
+  const [type, condition, dataSourceName] = watch(['type', 'condition', 'dataSourceName']);
+
+  const isGrafanaManagedType = type === RuleFormType.grafana;
+  const isCloudAlertRuleType = type === RuleFormType.cloudAlerting;
+  const isRecordingRuleType = type === RuleFormType.cloudRecording;
+
+  const showCloudExpressionEditor = (isRecordingRuleType || isCloudAlertRuleType) && dataSourceName;
+
+  const cancelQueries = useCallback(() => {
+    runner.current.cancel();
+  }, []);
+
+  const runQueries = useCallback(() => {
+    runner.current.run(queries);
+  }, [queries]);
+
+  // whenever we update the queries we have to update the form too
+  useEffect(() => {
+    setValue('queries', queries, { shouldValidate: false });
```

Comment:
I remember you mentioned this as an important sync step. Is there already/are you planning to add a test to cover this?